File: Problems/Blind75/ArrayHashing/two_sum.rs
// https://leetcode.com/problems/two-sum/
use std::collections::HashMap;

pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut nMap : HashMap<i32, i32> = HashMap::new();

    for i in 0..nums.len() {
        let rem : i32 = target - nums[i];

        if nMap.contains_key(&rem) {
            return vec![*nMap.get_mut(&rem).unwrap(), i as i32];
        }

        nMap.insert(nums[i].clone(), i as i32);
    }

    vec![]
}

File: Problems/Neetcode-150/arrays_hashing/two-sum.cpp
// https://leetcode.com/problems/two-sum/

class Solution {
public:
  vector<int> twoSum(vector<int> &nums, int target) {
    unordered_map<int, int> memo;
    for (int i = 0; i < nums.size(); i++) {
      const int x1 = nums[i];
      const int x2 = target - nums[i];

      if (memo.count(x2)) {
        return {i, memo[x2]};
      }

      memo[x1] = i;
    }

    return {-1, -1};
  }
};